pkg/replica/diff_disk.go

```diff
@@ -222,30 +222,42 @@ func (d *diffDisk) fullReadAt(buf []byte, offset int64) (int, error) {
 
 func (d *diffDisk) read(target byte, buf []byte, startOffset int64, startSector int64, sectors int64) (int, error) {
 	bufStart := startSector * d.sectorSize
+
 	bufLength := sectors * d.sectorSize
-	offset := startOffset + bufStart
-	size, err := d.files[target].Size()
+	diskReadStartOffset := startOffset + bufStart
+	diskReadEndOffset := diskReadStartOffset + bufLength
+	dSize, err := d.files[target].Size()
 	if err != nil {
 		return 0, err
 	}
+	vSize := d.size
 
-	// Reading the out-of-bound part is not allowed
-	if bufLength > d.size-offset {
-		logrus.Warn("Trying to read the out-of-bound part")
-		return 0, io.ErrUnexpectedEOF
-	}
+	count := 0
 
-	// May read the expanded part
-	if offset >= size {
-		return 0, nil
+	if diskReadStartOffset < dSize {
+		var newBuf []byte
+		if bufLength > dSize-diskReadStartOffset {
+			newBuf = buf[bufStart : bufStart+dSize-diskReadStartOffset]
+		} else {
+			newBuf = buf[bufStart : bufStart+bufLength]
+		}
+		n, err := d.files[target].ReadAt(newBuf, diskReadStartOffset)
+		if err != nil {
+			return n, err
+		}
+		count += n
 	}
-	var newBuf []byte
-	if bufLength > size-offset {
-		newBuf = buf[bufStart : bufStart+size-offset]
-	} else {
-		newBuf = buf[bufStart : bufStart+bufLength]
+	if diskReadEndOffset <= dSize {
+		return count, nil
 	}
-	return d.files[target].ReadAt(newBuf, offset)
+
+	// Read out of disk boundary but still in Volume boundary, we should remain the valid range with 0 and return the correct count
+	if diskReadEndOffset <= vSize {
+		return count + int(diskReadEndOffset-max(dSize, diskReadStartOffset)), nil
+	}
+
+	// Read out of Volume boundary, we should remain the valid range with 0 and return the correct count with EOF error
+	return count + int(vSize-max(dSize, diskReadStartOffset)), io.ErrUnexpectedEOF
```

Comment:
Will you create a e2e test for the EOF case?

Reply:
I will add a unit test for the function
how to handle EOF is the responsible of the upper application
so I don't think we need a full e2e test for EOF as long as we make sure we return EOF in correct cases

Reply:
One question, will a generic block device such a loop device or disk return the EOF in the same situation?
Just for making sure Longhorn disk can respond as same as a generic block device.

Reply:
From the golang io package
I think yes, it will return available data with count < len(buf) and EOF like what we do here
```
// ReaderAt is the interface that wraps the basic ReadAt method.
//
// ReadAt reads len(p) bytes into p starting at offset off in the
// underlying input source. It returns the number of bytes
// read (0 <= n <= len(p)) and any error encountered.
//
// When ReadAt returns n < len(p), it returns a non-nil error
// explaining why more bytes were not returned. In this respect,
// ReadAt is stricter than Read.
//
// Even if ReadAt returns n < len(p), it may use all of p as scratch
// space during the call. If some data is available but not len(p) bytes,
// ReadAt blocks until either all the data is available or an error occurs.
// In this respect ReadAt is different from Read.
//
// If the n = len(p) bytes returned by ReadAt are at the end of the
// input source, ReadAt may return either err == EOF or err == nil.
//
// If ReadAt is reading from an input source with a seek offset,
// ReadAt should not affect nor be affected by the underlying
// seek offset.
//
// Clients of ReadAt can execute parallel ReadAt calls on the
// same input source.
//
// Implementations must not retain p.
type ReaderAt interface {
	ReadAt(p []byte, off int64) (n int, err error)
}
```

Reply:
```
// ReaderAt is the interface that wraps the basic ReadAt method.
//
// ReadAt reads len(p) bytes into p starting at offset off in the
// underlying input source. It returns the number of bytes
// read (0 <= n <= len(p)) and any error encountered.
//
// When ReadAt returns n < len(p), it returns a non-nil error
// explaining why more bytes were not returned. In this respect,
// ReadAt is stricter than Read.
//
// Even if ReadAt returns n < len(p), it may use all of p as scratch
// space during the call. If some data is available but not len(p) bytes,
// ReadAt blocks until either all the data is available or an error occurs.
// In this respect ReadAt is different from Read.
//
// If the n = len(p) bytes returned by ReadAt are at the end of the
// input source, ReadAt may return either err == EOF or err == nil.
//
// If ReadAt is reading from an input source with a seek offset,
// ReadAt should not affect nor be affected by the underlying
// seek offset.
//
// Clients of ReadAt can execute parallel ReadAt calls on the
// same input source.
//
// Implementations must not retain p.
type ReaderAt interface {
	ReadAt(p []byte, off int64) (n int, err error)
}
```

Both Reader and ReadAt in golang all will return partial data with count <= len(buf) with non err

Reply:
Thanks for the confirmation. LGTM.

Reply:
In the end we use golang io ReadAt to read the file, so the way we handle EOF in our wrapper function follows the same semantics here
```
n, err := d.files[target].ReadAt(newBuf, diskReadStartOffset)
```